### PR TITLE
Convert existing commands that have flags to use the flag package and existing convention

### PR DESCRIPTION
**Pull request by sirnewton01**
Any commands that have command-line flags should now be using the flag package and following existing conventions. Some commands that did not have any flags, just arguments were preserved but hook into the existing help and usage conventions. Introduced a new multi-valued string flag var so to support flags that can be specified more than once where the order of the values could be important.

### CODE REVIEW

File: cmd/updateref.go [hunk 1]
```diff
@@ -72,6 +72,9 @@ func UpdateRef(c *git.Client, args []string) error {
 
 		return git.UpdateRef(c, opts, vals[0], cmt, *reason)
 	}
+
 	flags.Usage()
+	os.Exit(2)
+
```
[the remark this answers is not in the document]
Comment (sirnewton01) [reply cmd/updateref.go]: I don't think that the go compiler knows that os.Exit() will exit the process right away and so it needs a return after.. The return below is to satisfy the compiler and any code-path changes that allow it to be reached.

File: cmd/log.go [hunk 2]
```diff
@@ -67,9 +67,14 @@ func walkParents(c *git.Client, cmt git.CommitID) error {
...
+	if len(args) == 1 && args[0] == "--help" {
+		flag.Usage()
+		os.Exit(0)
+	}
+
 	if len(args) >= 2 {
-		fmt.Fprintf(os.Stderr, "Usage: %s log [commitish]\n", os.Args[0])
-		return errors.New("No options are currently supported for log")
+		flag.Usage()
+		os.Exit(2)
```
[the remark this answers is not in the document]
Comment (sirnewton01) [reply cmd/log.go]: It's a tricky problem. Should we be describing our usage based on what the official git does or based on what we have implemented? The flag package uses the declared usage flags and is going to report its own errors when a flag that we haven't implemented yet is used by an end user and it looks much much like this case here.

File: cmd/diffindex.go [hunk 3]
```diff
@@ -25,7 +26,9 @@ func DiffIndex(c *git.Client, args []string) error {
 	}
 
 	if len(args) < 1 {
-		return fmt.Errorf("Must provide a treeish to git diff-index")
+		fmt.Fprintf(flag.CommandLine.Output(), "Must provide a treeish to git diff-index\n")
+		flags.Usage()
+		os.Exit(2)
```
[the remark this answers is not in the document]
Comment (sirnewton01) [reply cmd/diffindex.go]: I'm trying to make usage failures uniform. These three lines will simulate what the flag package will do if there is some problem with the flags, but in other usage scenarios, such as with the arguments. So, it prints a message to the flag's current output writer (stdout, or stderr depending on the case), outputs the usage to the same writer under the hood and exit code 2. Flag will do the same thing if you do something like 'git status --foobar' except it produces its own messages.

File: cmd/commit.go [hunk 4]
```diff
@@ -32,48 +32,57 @@ func Commit(c *git.Client, args []string) (string, error) {
...
+			return "", err
 		}
+		message = append(message, string(f))
 	}
+
+	if len(message) == 0 || edit {
+		opts.NoEdit = false
+	}
+
+	finalMessage := strings.Join(message, "\n")
```
[the remark this answers is not in the document]
Comment (sirnewton01) [reply cmd/commit.go]: Thanks, I'll look into this.

**Comment by sirnewton01** [comment]
@driusan yes, I've tried the whole make t0*.sh from the official git in order to try to find any regressions. I do a '| grep -e '^ok' | wc -l to find the number of passing tests before and after.

File: cmd/updateref.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (driusan) [reply cmd/updateref.go]: Makes sense, I didn't consider that.

File: cmd/log.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (driusan) [reply cmd/log.go]: I think we should generally try to parse all the options that are listed in the man page, even if we don't support them, to ensure command line compatibility.. then an error message saying something like "--foo not implemented" is obvious.

This isn't really feasible for things like diff that have 6 million options, but for most plumbing commands it should be.

File: cmd/diffindex.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (driusan) [reply cmd/diffindex.go]: Okay, I'm not strongly opposed since cmd is basically an extension of package main in this case.

File: .travis.yml [hunk 5]
```diff
@@ -4,9 +4,12 @@ go:
...
 
+go_import_path: github.com/driusan/dgit
+
 script:
+    - pwd
     - diff -u <(echo -n) <(gofmt -d ./)
     - go test -v ./...
     - chmod u+x run-official-tests.sh
-    - ./run-official-tests.sh t0000-basic.sh
+    - ./run-official-tests.sh -i t0000-basic.sh t0011-hashmap.sh t0040-parse-options.sh t0062-revision-walking.sh t0063-string-list.sh t0064-sha1-array.sh t0070-fundamental.sh t0081-line-buffer.sh t0110-urlmatch-normalization.sh t0200-gettext-basic.sh
```
Comment (driusan) [inline .travis.yml]: t0011-hashmap.sh, t0040-parse-options.sh, t0063-string-list.sh, t0064-sha1-array.sh, t0110-urlmatch-normalization.sh and t0200-gettext-basic.sh don't seem to ever invoke dgit. I think this is mostly testing test-tool, which is built by git.

I think the most important ones (which are currently failing, but we currently allow it to fail) are the -m and -u read-tree tests. Maybe some of the update-index ones. Basically, things that can corrupt the index. (I'd say write-tree was even more important, but I don't see any tests specifically testing write-tree.)

File: main.go [hunk 6]
```diff
@@ -313,6 +328,12 @@ func main() {
 			fmt.Fprintln(os.Stderr, err)
 			os.Exit(4)
 		}
+	case "var":
```
Comment (driusan) [inline main.go]: I think you committed this to the wrong branch? Is this for a different PR?

Comment (sirnewton01) [reply main.go]: Thanks, I'll defer this until a future PR.

File: .travis.yml (hunk 5 above)
Comment (sirnewton01) [reply .travis.yml]: I was adding the suites that seem to be passing fine in the hopes of eventually just running all of them. I'll revert back to the ones that we think are actually testing dgit.

Comment (driusan) [reply .travis.yml]: We could consider writing a test-tool replacement which tests *our* internals in the same way and hiding it behind a build flag (I haven't looked into how complicated that would be), but until running the tests fails the build I don't see much point in running the tests which aren't testing our functionality.

**Pull request merged**